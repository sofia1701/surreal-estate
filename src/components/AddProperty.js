import React, { useState } from "react";
import "../styles/add-property.css";

export default function AddProperty() {
  const initialState = {
    fields: {
      title: "",
      city: "Manchester",
      type: "Flat",
    },
  };

  const [fields, setFields] = useState(initialState.fields);

  const handleAddProperty = (event) => {
    event.preventDefault();
    console.log(fields);
  };

  const handleFieldChange = (event) => {
    setFields({ ...fields, [event.target.name]: event.target.value });
  };

  return (
    <div className="add-property">
      <form
        data-testid="form-id"
        className="form-add-property"
        onSubmit={handleAddProperty}
      >
        <label htmlFor="title">
          Title
          <input
            className="input-fields"
            placeholder="Add a title..."
            id="title"
            name="title"
            value={fields.title}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="city">
          City
          <select
            data-testid="select-city-id"
            className="input-fields"
            id="city"
            name="city"
            value={fields.city}
            onChange={handleFieldChange}
          >
            <option value="Manchester">Manchester</option>
            <option value="Leeds">Leeds</option>
            <option value="Sheffield">Sheffield</option>
            <option value="Liverpool">Liverpool</option>
          </select>
        </label>
        <label htmlFor="type">
          Type
          <select
            data-testid="select-type-id"
            className="input-fields"
            id="type"
            name="type"
            value={fields.type}
            onChange={handleFieldChange}
          >
            <option value="Flat">Flat</option>
            <option value="Detached">Detached</option>
            <option value="Semi-Detached">Semi-Detached</option>
            <option value="Terraced">Terraced</option>
            <option value="End of Terraced">End of Terraced</option>
            <option value="Cottage">Cottage</option>
            <option value="Bungalow">Bungalow</option>
          </select>
        </label>
        <button className="add-button" type="submit">
          Add
        </button>
      </form>
    </div>
  );
}
