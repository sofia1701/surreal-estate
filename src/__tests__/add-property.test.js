import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import AddProperty from "../components/AddProperty";

describe("AddProperty", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("captures user input correctly", () => {
    const { getByRole } = render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const textbox = getByRole("textbox");
    const button = getByRole("button", { name: /add/i });

    fireEvent.change(textbox, {
      target: { value: "3 bedroom flat" },
    });
    fireEvent.click(button);

    expect(textbox.value).toBe("3 bedroom flat");
  });

  it("captures user select city correctly", () => {
    const { getByTestId, getByText } = render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );
    expect(getByTestId("form-id")).toHaveFormValues({ city: "Manchester" });

    fireEvent.change(getByTestId("select-city-id"), {
      target: { value: "Liverpool" },
    });
    expect(getByText("Liverpool")).toBeInTheDocument();
    expect(getByTestId("form-id")).toHaveFormValues({ city: "Liverpool" });
  });

  it("captures user select type correctly", () => {
    const { getByTestId, getByText } = render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );
    expect(getByTestId("form-id")).toHaveFormValues({ type: "Flat" });

    fireEvent.change(getByTestId("select-type-id"), {
      target: { value: "Detached" },
    });
    expect(getByText("Detached")).toBeInTheDocument();
    expect(getByTestId("form-id")).toHaveFormValues({ type: "Detached" });
  });
});
