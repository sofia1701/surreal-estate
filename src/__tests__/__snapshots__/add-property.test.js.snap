// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProperty renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-property"
  >
    <form
      class="form-add-property"
      data-testid="form-id"
    >
      <label
        for="title"
      >
        Title
        <input
          class="input-fields"
          id="title"
          name="title"
          placeholder="Add a title..."
          value=""
        />
      </label>
      <label
        for="city"
      >
        City
        <select
          class="input-fields"
          data-testid="select-city-id"
          id="city"
          name="city"
        >
          <option
            value="Manchester"
          >
            Manchester
          </option>
          <option
            value="Leeds"
          >
            Leeds
          </option>
          <option
            value="Sheffield"
          >
            Sheffield
          </option>
          <option
            value="Liverpool"
          >
            Liverpool
          </option>
        </select>
      </label>
      <label
        for="type"
      >
        Type
        <select
          class="input-fields"
          data-testid="select-type-id"
          id="type"
          name="type"
        >
          <option
            value="Flat"
          >
            Flat
          </option>
          <option
            value="Detached"
          >
            Detached
          </option>
          <option
            value="Semi-Detached"
          >
            Semi-Detached
          </option>
          <option
            value="Terraced"
          >
            Terraced
          </option>
          <option
            value="End of Terraced"
          >
            End of Terraced
          </option>
          <option
            value="Cottage"
          >
            Cottage
          </option>
          <option
            value="Bungalow"
          >
            Bungalow
          </option>
        </select>
      </label>
      <button
        class="add-button"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
